fncas/fncas_optimize.h

```diff
@@ -0,0 +1,266 @@
+    while (test_f_value > (current_f_value - bt_alpha_ * t * squared_gradient)) {
+      t *= bt_beta_;
+      for (size_t i = 0; i < dim; ++i) {
+        test_point[i] = current_point[i] - t * current_f_gradf.gradient[i];
+      }
+      test_f_value = fi(test_point);
+      if (bt_iteration++ == bt_max_steps_) {
+        break;
+      }
+    }
+    current_point = test_point;
+
+    // Simple early stopping by the norm of gradient.
+    if (std::sqrt(squared_gradient) < grad_eps_) {
+      break;
+    }
+  }
+
+ return OptimizationResult{fi(current_point), current_point};
+}
+
+template<class F>
+double ConjugateGradientOptimizer<F>::PolakRibiere(const std::vector<double> &g,
+                                                   const std::vector<double> &g_prev) {
+  double beta = (fncas::L2Norm(g) - fncas::DotProduct(g, g_prev)) / fncas::L2Norm(g_prev);
+  if (!std::isinf(beta) && !std::isnan(beta)) {
+    return std::max(0.0, beta);
+  } else {
+    return 0.0;
+  }
+}
+
+// TODO(mazhurovich): Implement more sophisticated version.
+template<class F> 
+OptimizationResult ConjugateGradientOptimizer<F>::Optimize(const std::vector<double>& starting_point) {
+  fncas::reset_internals_singleton();
+  const size_t dim = starting_point.size();
+  const fncas::x gradient_helper(dim);
+  fncas::f_intermediate fi(F::compute(gradient_helper));
+  fncas::g_intermediate gi = fncas::g_intermediate(gradient_helper, fi);
+  std::vector<double> current_point(starting_point);
+
+  const auto initial_f_gradf = gi(current_point);
+  double current_f_value = initial_f_gradf.value;
+  std::vector<double> current_grad = initial_f_gradf.gradient;
+  std::vector<double> s(current_grad);
+  FlipSign(s);
+
+  for (size_t iteration = 0; iteration < max_steps_; ++iteration) {
+    double squared_gradient = 0.0;
+    std::vector<double> test_point(dim);
+
+    for (size_t i = 0; i < dim; ++i) {
+      test_point[i] = current_point[i] + s[i];
+      squared_gradient += current_grad[i] * current_grad[i];
+    }
+
+    double t = 1.0;
+    double test_f_value = fi(test_point);
+    
+    // Naive backtracking with strict limit on the number of steps.
+    size_t bt_iteration = 0;
+    while (test_f_value > (current_f_value - bt_alpha_ * t * squared_gradient)) {
+      t *= bt_beta_;
+      for (size_t i = 0; i < dim; ++i) {
+        test_point[i] = current_point[i] + t * s[i];
+      }
+      test_f_value = fi(test_point);
+      if (bt_iteration++ == bt_max_steps_) {
+        break;
+      }
+    }
+
+    const auto new_f_gradf = gi(test_point);
+    const std::vector<double> new_grad = new_f_gradf.gradient;
+    const double omega = PolakRibiere(new_grad, current_grad);
+    for (size_t i = 0; i < dim; ++i) {
+      s[i] = omega * s[i] - new_grad[i];
+    }
+    current_point = test_point;
+    current_grad = new_grad;
+    current_f_value = new_f_gradf.value;
+
+    // Simple early stopping by the norm of gradient.
+    if (std::sqrt(L2Norm(s)) < grad_eps_) {
```

Comment:
Nit: For this early stopping criteria, here and above, I'd add the minimum number of steps before it happens as well.